torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp

```diff
@@ -0,0 +1,222 @@
+
+  if (typesFromScalars.size() == n->inputs().size()) {
+    // If all inputs are scalars, infer scalar_type by calling c10::promoteTypes.
+    st = PromoteScalarTypes(typesFromScalars);
+  } else if (output_st && !IsComparisonOp(n->kind())) {
+    // If output scalar type is available, use that.
+    st = output_st;
+  } else if (!typesFromTensors.empty()) {
+    // When inputs consist of tensors and scalars. In PyTorch, scalars are implicitly casted to have the
+    // same scalar type as input tensors.
+    st = typesFromTensors[0];
+    if (std::any_of(typesFromTensors.begin(), typesFromTensors.end(), [&st](const c10::ScalarType& type) {
+      return type != st;
+    })) {
+      std::cerr << "Warning: ONNX Scalar Type Analysis - Scalar types mismatch for tensor inputs of operator "
+                << n->kind().toDisplayString()
+                << ". Please report a bug to PyTorch. "
+                << "The scalar type "
+                << c10::toString(*st)
+                << " of the first tensor is chosen." << std::endl;
+    }
+  } else {
+    // When inputs consist of only scalars.
+    st = PromoteScalarTypes(typesFromScalars);
+  }
+
+  return st;
+}
+
+static void UpdateScalarTypeForInputs(Node* n, const c10::ScalarType& scalar_type) {
+  const int64_t onnx_type = ScalarTypeToONNXType(scalar_type);
+  if (onnx_type < 0) {
+    std::cerr << "Warning: ONNX Scalar Type Analysis - Scalar type: "
+              << c10::toString(scalar_type)
+              << " of input tensor in operator: " << n->kind().toDisplayString()
+              << " not supported in ONNX. " << std::endl;
+    return;
+  }
+
+  for (auto input : n->inputs()) {
+    auto input_tensor_type = input->type()->cast<TensorType>();
+    auto input_scalar_type = input_tensor_type->scalarType();
+
+    if ((input->node()->kind() == onnx::Constant) ||
+        (input_scalar_type && (*input_scalar_type != scalar_type))) {
+      if (input->node()->kind() == onnx::Constant) {
+        // Fix up the scalar directly instead of inserting a cast operator.
+        // NOTE: Keep only the else branch once constant_folding is enabled by
+        // default.
+        at::Tensor val = input->node()->t(attr::value);
+        at::Tensor new_val = val.to(scalar_type);
+        Node* const_node = n->owningGraph()->create(onnx::Constant);
+        const_node->t_(attr::value, new_val);
+        const_node->insertBefore(n);
+        const_node->output()->setType(TensorType::create(new_val));
+        n->replaceInputWith(input, const_node->output());
+      } else {
+        Node* cast_node = n->owningGraph()->create(onnx::Cast);
+        cast_node->addInput(input);
+        cast_node->i_(attr::to, onnx_type);
+        cast_node->insertBefore(n);
+        cast_node->output()->setType(CreateProfiledTensorTypeWithScalarType(
+            input_tensor_type, scalar_type));
+        n->replaceInputWith(input, cast_node->output());
```

Comment:
I wonder if this works correctly if a node has the same value as input several times, e.g
```
%a = prim::mul(%b, %b)
```
I'm not sure that `inputs()` iterator will stay valid after we call `replaceInputWith`, which would replace all entries of the current input. Could you please double-check that it works as expected? A safe alternative to achieve the same and avoid iterator issues for certain would be to use `replaceInput(index, newValue)`.

Reply:
That's a very interesting case. I will create a test case for that, after fixing the regression #26328.